deblur/workflow.py

```diff
@@ -41,9 +47,95 @@ def dereplicate_seqs(seqs_fp,
         log_name=log_name)
 
 
-def remove_artifacts_seqs(seqs_fp):
-    """Step 3: remove artifacts from FASTA file"""
-    pass
+def remove_artifacts_seqs(seqs_fp,
+                          ref_fp,
+                          output_fp,
+                          ref_db_fp=None,
+                          negate=False,
+                          threads=1):
+    """Step 3: remove artifacts from FASTA file using
+       SortMeRNA
+
+    Parameters
+    ----------
+    seqs_fp: string
+        file path to FASTA input sequence file
+    ref_fp: tuple
+        file path(s) to FASTA database file
+    output_fp: string
+        file path to store output results
+    ref_db_fp: string or tuple, optional
+        file path(s) to indexed FASTA database
+    negate: boolean, optional
+        if True, discard all input sequences aligning
+        to reference database
+    threads: integer, optional
+        number of threads to use for SortMeRNA
+    """
+    working_dir = join(dirname(output_fp), "working_dir")
+    if not exists(working_dir):
+        makedirs(working_dir)
+
+    aligned_seq_ids = set()
+    files_to_remove = []
+
+    for i, db in enumerate(ref_fp):
+        # create working directory for each
+        # reference database
+        db_dir_base = splitext(basename(db))[0]
+        db_dir = join(working_dir, db_dir_base)
+        if not exists(db_dir):
+            makedirs(db_dir)
+
+        if ref_db_fp:
+            sortmerna_db = ref_db_fp[i]
+        else:
+            # build index
+            sortmerna_db, files_to_remove = \
+                build_database_sortmerna(
+                    fasta_path=db,
+                    max_pos=10000,
+                    output_dir=db_dir)
+
+        # run SortMeRNA
+        app_result = sortmerna_map(
+            seq_path=seqs_fp,
+            output_dir=db_dir,
+            refseqs_fp=db,
+            sortmerna_db=sortmerna_db,
+            threads=threads,
+            best=1)
+
+        # Print SortMeRNA errors
+        stderr_fp = app_result['StdErr'].name
+        if stat(stderr_fp).st_size != 0:
+            with open(stderr_fp, 'U') as stderr_f:
+                for line in stderr_f:
+                    print line
+            raise ValueError("Could not run SortMeRNA.")
+
+        for line in app_result['BlastAlignments']:
+            line = line.strip().split('\t')
+            if line[1] == '*':
+                continue
+            else:
+                aligned_seq_ids.add(line[0])
+
+        # remove indexed database files
+        remove_files(files_to_remove, error_on_missing=False)
+
+    # if negate = False, only output sequences
+    # matching to at least one of the databases
+    with open(seqs_fp, 'U') as seqs_f:
```

Comment:
The `if negate` statement will always evaluate the same in a single execution of the function. You can rewrite this entire portion of code as follows in order to reduce running time and code duplication:

```python
if negate:
    op = lambda x: x not in aligned_seq_ids
else:
    op = lambda x: x in aligned_seq_ids

with open(seqs_fp, 'U') as seqs_f:
    with open(output_fp, 'w') as out_f:
        for label, seq in parse_fasta(seqs_f):
            label = label.split()[0]
            if op(label):
                out_f.write(">%s\n%s\n" % (label, seq))
```